ckan/lib/files/__init__.py

```diff
@@ -0,0 +1,216 @@
+    }
+
+    for plugin in p.PluginImplementations(p.IFiles):
+        result.update(plugin.files_get_storage_adapters())
+
+    return result
+
+
+def collect_storage_configuration(
+    config: Mapping[str, Any], prefix: str = STORAGE_PREFIX, /, flat: bool = False
+):
+    """Return settings of every storage located in the config.
+
+    Args:
+        config: mapping with configuration
+        prefix: common prefix for storage options
+
+    Keyword Args:
+        flat: do not transfrom nested keys into dictionaries
+
+    Returns:
+        dictionary with configuration of all storages
+
+    """
+    storages = defaultdict(dict)  # type: dict[str, dict[str, Any]]
+    prefix_len = len(prefix)
+
+    # first, group config options by the storage name
+    for k, v in config.items():
+        if not k.startswith(prefix):
+            continue
+
+        # when `flat` flag is enabled, nested keys `a.b.c` kept as a string:
+        # `{"a.b.c": ...}`. When it's disabled, transform such keys into nested
+        # dictionaries `{"a": {"b": {"c": ...}}}`
+        name, *path = k[prefix_len:].split(".", 1 if flat else -1)
+        if not path:
+            log.warning("Unrecognized storage configuration: %s = %s", k, v)
+            continue
+
+        here = storages[name]
+        for segment in path[:-1]:
+            here = here.setdefault(segment, {})
+
+        here[path[-1]] = v
+
+    return storages
+
+
+def collect_storages() -> dict[str, Storage]:
+    """Initialize configured storages.
+
+    Returns:
+        mapping with storages
+    """
+    result = {}
+
+    mapping = collect_storage_configuration(config)
+
+    for name, settings in mapping.items():
+        try:
+            storage = make_storage(name, settings)
+        except (
+            exc.UnknownAdapterError,
+            exc.InvalidStorageConfigurationError,
+        ) as err:
+            raise CkanConfigurationException(str(err)) from err
+
+        storages.register(name, storage)
+
+    if path := config["ckan.storage_path"]:
+        result["resources"] = make_storage(
+            "resources",
+            {
+                "type": "ckan:fs",
+                "path": os.path.join(path, "resources"),
+                "create_path": True,
+                "recursive": True,
+                "override_existing": True,
+                "location_transformers": ["safe_relative_path"],
+                "max_size": config["ckan.max_resource_size"] * 1024 * 1024,
+            },
+        )
+
+        for object_type in ["user", "group", "admin"]:
+            name = f"{object_type}_uploads"
+            result[name] = make_storage(
+                name,
+                {
+                    "type": "ckan:fs",
+                    "path": os.path.join(path, "storage", "uploads", object_type),
+                    "create_path": True,
+                    "max_size": config["ckan.max_image_size"] * 1024 * 1024,
+                },
+            )
+
+    return result
```

Comment:
Shouldn't `result` include the storages created from the config options (registered with `storages.register()`), and conversely, the `resources`, `user` etc storages also registered with `storages.register()`)?

Basically it looks like the "new" and legacy storages are not tracked in the same registry.

Reply:
I was rewriting this function multiple times and didn't notice this discrepancy. In this PR, I'll update online to register everything in the same dictionary, but `resources`, `user`, ... will override configured storages with the same name. 
In the second phase, this function is further improved, and it will check the existing configuration before overriding storage with hardcoded versions.